pkg/virt-controller/watch/clone/clone.go

```diff
@@ -83,193 +89,194 @@ func (ctrl *VMCloneController) execute(key string) error {
 	case sourceTypeVM:
-		vmKey := getKey(sourceInfo.Name, vmClone.Namespace)
-		obj, vmExists, err := ctrl.vmInformer.GetStore().GetByKey(vmKey)
+		sourceVMObj, err := ctrl.getSource(vmClone, sourceInfo.Name, vmClone.Namespace, string(sourceTypeVM), ctrl.vmInformer.GetStore())
 		if err != nil {
-			return fmt.Errorf("error getting VM %s in namespace %s from cache: %v", sourceInfo.Name, vmClone.Namespace, err)
+			return syncInfo, err
 		}
-		if !vmExists {
-			err = ctrl.updateStatus(vmClone, syncInfoType{
-				isCloneFailing: true,
-				failEvent:      SnapshotNotCreated,
-				failReason:     fmt.Sprintf("VirtualMachine %s does not exist in namespace %s", vmClone.Spec.Source.Name, vmClone.Namespace),
-			})
-
-			if err != nil {
-				log.Log.Errorf("updating status when source vm does not exist failed: %v", err)
-			}
 
-			return fmt.Errorf("VM %s in namespace %s does not exist", sourceInfo.Name, vmClone.Namespace)
+		sourceVM := sourceVMObj.(*k6tv1.VirtualMachine)
+
+		syncInfo = ctrl.syncSourceVM(sourceVM, vmClone)
+		return syncInfo, nil
+
+	case sourceTypeSnapshot:
+		sourceSnapshotObj, err := ctrl.getSource(vmClone, sourceInfo.Name, vmClone.Namespace, string(sourceTypeSnapshot), ctrl.snapshotInformer.GetStore())
+		if err != nil {
+			return syncInfo, err
 		}
-		sourceVM := obj.(*k6tv1.VirtualMachine)
 
-		syncInfo = ctrl.syncSourceVM(key, sourceVM, vmClone)
-		syncErr = syncInfo.err
+		sourceSnapshot := sourceSnapshotObj.(*snapshotv1alpha1.VirtualMachineSnapshot)
+
+		syncInfo = ctrl.syncSourceSnapshot(sourceSnapshot, vmClone)
+		return syncInfo, nil
+
 	default:
-		return fmt.Errorf("clone %s is defined with an unknown source type %s", vmClone.Name, sourceInfo.Kind)
+		return syncInfo, fmt.Errorf("clone %s is defined with an unknown source type %s", vmClone.Name, sourceInfo.Kind)
 	}
+}
 
-	err = ctrl.updateStatus(vmClone, syncInfo)
-	if err != nil {
-		return fmt.Errorf("error updating status: %v", err)
-	}
+func (ctrl *VMCloneController) syncSourceVM(source *k6tv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
+	targetType := ctrl.getTargetType(vmClone)
 
-	if syncErr != nil {
-		return fmt.Errorf("sync error: %v", syncErr)
-	}
+	switch targetType {
+	case targetTypeVM:
+		return ctrl.syncSourceVMTargetVM(source, vmClone)
 
-	return nil
+	default:
+		return syncInfoType{err: fmt.Errorf("target type is unknown: %s", targetType)}
+	}
 }
 
-func (ctrl *VMCloneController) syncSourceVM(key string, source *k6tv1.VirtualMachine, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
-	var targetType cloneTargetType
-	if vmClone.Spec.Target != nil {
-		targetType = cloneTargetType(vmClone.Spec.Target.Kind)
-	} else {
-		targetType = defaultType
-	}
+func (ctrl *VMCloneController) syncSourceSnapshot(source *snapshotv1alpha1.VirtualMachineSnapshot, vmClone *clonev1alpha1.VirtualMachineClone) syncInfoType {
+	targetType := ctrl.getTargetType(vmClone)
 
 	switch targetType {
 	case targetTypeVM:
```

Comment:
That's a great question :D
Obviously an answer demands some more discussion and thought, but this what comes into mind:
- VMI
- VM pools
- Snapshot
  - This may be helpful to "modify" a snapshot (i.e. create a similar snapshot and possibly delete the original one). For instance, we can change VM's spec / labels / whatever (using clone's spec), then delete the old snapshot. In the future more spec fields would be added to clones, so we could probably also add / remove / wipe disks.

I'm sure more ideas would pop up in the future :)